parser/transaction.go

```diff
@@ -397,76 +381,193 @@ func (tx *Transaction) ParseFromSlice(data []byte) ([]byte, error) {
+		tx.shieldedSpends[i] = newSpend
+	}
+	if !s.ReadCompactSize(&outputCount) {
+		return nil, errors.New("could not read nShieldedOutput")
+	}
+	tx.shieldedOutputs = make([]*output, outputCount)
+	for i := 0; i < outputCount; i++ {
+		newOutput := &output{}
+		s, err = newOutput.ParseFromSlice([]byte(s), tx.version)
+		if err != nil {
+			return nil, errors.Wrap(err, "while parsing shielded Output")
+		}
+		tx.shieldedOutputs[i] = newOutput
+	}
+	if spendCount+outputCount > 0 && !s.ReadInt64(&tx.valueBalanceSapling) {
+		return nil, errors.New("could not read valueBalance")
+	}
+	if spendCount > 0 && !s.Skip(32) {
+		return nil, errors.New("could not skip anchorSapling")
+	}
+	if !s.Skip(192 * spendCount) {
+		return nil, errors.New("could not skip vSpendProofsSapling")
+	}
+	if !s.Skip(64 * spendCount) {
+		return nil, errors.New("could not skip vSpendAuthSigsSapling")
+	}
+	if !s.Skip(192 * outputCount) {
+		return nil, errors.New("could not skip vOutputProofsSapling")
+	}
+	if spendCount+outputCount > 0 && !s.ReadBytes(&tx.bindingSigSapling, 64) {
+		return nil, errors.New("could not read bindingSigSapling")
+	}
+	var actionsCount, proofsCount int
+	if !s.ReadCompactSize(&actionsCount) {
+		return nil, errors.New("could not read nActionsOrchard")
+	}
+	if !s.Skip(820 * actionsCount) {
+		return nil, errors.New("could not skip vActionsOrchard")
+	}
+	if actionsCount > 0 {
+		if !s.Skip(1) {
+			return nil, errors.New("could not skip flagsOrchard")
+		}
+		if !s.Skip(8) {
+			return nil, errors.New("could not skip valueBalanceOrchard")
+		}
+		if !s.Skip(32) {
+			return nil, errors.New("could not skip anchorOrchard")
+		}
+		if !s.ReadCompactSize(&proofsCount) {
+			return nil, errors.New("could not read sizeProofsOrchard")
+		}
+		if !s.Skip(proofsCount) {
+			return nil, errors.New("could not skip proofsOrchard")
+		}
+		if !s.Skip(64 * actionsCount) {
+			return nil, errors.New("could not skip vSpendAuthSigsOrchard")
+		}
+		if !s.Skip(64) {
+			return nil, errors.New("could not skip bindingSigOrchard")
 		}
 	}
+	return s, nil
+}
+
+// ParseFromSlice deserializes a single transaction from the given data.
+func (tx *Transaction) ParseFromSlice(data []byte) ([]byte, error) {
+	s := bytestring.String(data)
+
+	// declare here to prevent shadowing problems in cryptobyte assignments
+	var err error
+
+	var header uint32
+	if !s.ReadUint32(&header) {
+		return nil, errors.New("could not read header")
+	}
+
+	tx.fOverwintered = (header >> 31) == 1
+	tx.version = header & 0x7FFFFFFF
 
+	if tx.version >= 3 {
+		if !s.ReadUint32(&tx.nVersionGroupID) {
+			return nil, errors.New("could not read nVersionGroupId")
+		}
+	}
+	// parse the main part of the transaction
+	if tx.version <= 4 {
+		s, err = tx.parseV234([]byte(s), tx.version)
+	} else {
+		s, err = tx.parseV5([]byte(s), tx.version)
+	}
+	if err != nil {
+		return nil, err
+	}
 	// TODO: implement rawBytes with MarshalBinary() instead
 	txLen := len(data) - len(s)
 	tx.rawBytes = data[:txLen]
```

Comment:
Oh, I guess I see - the `data` buffer may exceed the size of the transaction? So the check should be that the remainder of the buffer is zeroed out or something?